### PR TITLE
Alternative proposal for breaking change to the way plugins load cli options

### PR DESCRIPTION
**Pull request by antonydenyer**
https://wiki.hyperledger.org/display/BESU/DRAFT+-+Pico+CLI+Plugin+Integration

The main change is to move preparePlugins into the begging of the run method. This means that all the cli parsing will happen before the method register is called. 

The only way to register cli options is to use the extra hook provided by `BesuPlugins::registerPicoCLIOptions`

### CODE REVIEW

File: plugin-api/src/main/java/org/hyperledger/besu/plugin/BesuPlugin.java [hunk 1]
```diff
@@ -36,6 +38,16 @@
     return Optional.of(this.getClass().getName());
   }
 
+  /**
+   * Called when the plugin is loaded so that you can register cli options to be parsed
+   *
+   * @param cliOptions the hook to register PicoCLIOptions.
+   */
+  default void registerPicoCLIOptions(final PicoCLIOptions cliOptions) {
```
Comment (shemnon) [inline plugin-api/src/main/java/org/hyperledger/besu/plugin/BesuPlugin.java:46]: My concern is that, even with a default interface method, it forces all plugins to be PicoCLI aware and deeply embedded in the lifecycle.  Not every plugin cares or should care about PicoCLI, which is why I exposed it as a service.

Comment (antonydenyer) [reply plugin-api/src/main/java/org/hyperledger/besu/plugin/BesuPlugin.java:46]: Yeah, I'm not a fan of this either. However, it is more explicit than the other proposal https://github.com/hyperledger/besu/pull/2768.

**Comment by jflo** [comment]
Is this still active or can it be closed? We would like to clean up old outstanding PRs.

**Comment by antonydenyer** [comment]
Related to https://wiki.hyperledger.org/display/BESU/DRAFT+-+Pico+CLI+Plugin+Integration and https://github.com/hyperledger/besu/pull/2768

Maybe useful if someone looks at plugins again.

**Pull request closed without merging**